samcli/lib/utils/git_repo.py

```diff
@@ -0,0 +1,188 @@
+            when reaching unstable state, for example with replace_existing flag set, unstable state can happen
+            if removed the current local clone but failed to copy the new one from the temp location to the destination
+        """
+        if not self._auto_clone:  # Unit test escape hatch
+            return None  # type: ignore
+        GitRepo._ensure_clone_directory_exists(clone_dir=clone_dir)
+        # clone to temp then move to the destination(repo_local_path)
+        with osutils.mkdir_temp(ignore_errors=True) as tempdir:
+            try:
+                clone_name = clone_name if clone_name else self._name
+                temp_path = os.path.normpath(os.path.join(tempdir, clone_name))
+                git_executable: str = GitRepo._git_executable()
+                LOG.info("\nCloning from %s", self._url)
+                subprocess.check_output(
+                    [git_executable, "clone", self._url, clone_name],
+                    cwd=tempdir,
+                    stderr=subprocess.STDOUT,
+                )
+                self._local_path = self._persist_local_repo(temp_path, clone_dir, clone_name, replace_existing)
+                return self._local_path
+            except OSError as ex:
+                LOG.warning("WARN: Could not clone repo %s", self._url, exc_info=ex)
+                raise
+            except subprocess.CalledProcessError as clone_error:
+                output = clone_error.output.decode("utf-8")
+                if "not found" in output.lower():
+                    LOG.warning("WARN: Could not clone repo %s", self._url, exc_info=clone_error)
+                raise CloneRepoException from clone_error
+            finally:
+                self._clone_attempted = True
+
+    @staticmethod
+    def _persist_local_repo(temp_path: str, dest_dir: Path, dest_name: str, replace_existing: bool) -> Path:
+        dest_path = os.path.normpath(dest_dir.joinpath(dest_name))
+        try:
+            if Path(dest_path).exists():
+                if not replace_existing:
+                    raise CloneRepoException(f"Can not clone to {dest_path}, directory already exist")
+                LOG.debug("Removing old repo at %s", dest_path)
+                shutil.rmtree(dest_path, onerror=rmtree_callback)
+
+            LOG.debug("Copying from %s to %s", temp_path, dest_path)
+            shutil.copytree(temp_path, dest_path, ignore=shutil.ignore_patterns("*.git"))
+            return Path(dest_path)
+        except (OSError, shutil.Error) as ex:
+            # UNSTABLE STATE
+            # it's difficult to see how this scenario could happen except weird permissions, user will need to debug
+            raise CloneRepoUnstableStateException(
+                "Unstable state when updating repo. "
+                f"Check that you have permissions to create/delete files in {dest_dir} directory "
+                "or file an issue at https://github.com/awslabs/aws-sam-cli/issues"
```

Comment:
We may want to remove this line as we are not going to put it into all our error messages

Reply:
This is the same as before, I didn't change it.

Reply:
This one is re-raised as a UserException (click.ClickException) and the message is displayed to the user